src/emulator/storage/rules/runtime.ts

```diff
@@ -455,6 +457,31 @@ async function fetchFirestoreDocument(
   }
 }
 
+/**
+ * Deeply replaces objects matching {nullValue: null} with {nullValue: 0}.
+ * This corresponds to how Rules emulator handles the null_value type (as an
```

Comment:
This doesn't seem right. To quote the protobuf reference docs:

> The JSON representation for `NullValue` is JSON `null`.

https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#nullvalue

This should be special-cased by the Java protobuf library already. In other words, if the Rules emulator doesn't accept it, that's a bug within the Rules emulator and should be fixed there. Client side fixes like this may mask bigger problems.